draft-ietf-quic-transport.md

```diff
@@ -3007,25 +3007,26 @@ guidance offered below seeks to strike this balance.
 
 ### Sending ACK Frames {#sending-acknowledgements}
 
+Every packet SHOULD be acknowledged at least once, and ack-eliciting packets
+MUST be acknowledged at least once within the maximum ack delay. An endpoint
```

Comment:
I think I agree with @mikkelfj that MUST might not be the right word here. It's obviously unenforceable for one. But also, I think there are valid reasons where a packet won't be acknowledged. For one, we have a memory limit for the data structure we use to keep track of packets that need to be acknowledged. At a high enough rate, with enough (random?) packet loss you could have enough gaps to hit that limit. In that case, we start forgetting older packet numbers, and will never end up acknowledging them.

Reply:
We have a memory limit too, but that doesn't mean that we don't send acknowledgments.  We just send packets more often as a result.

As for untestable, that's true for a bunch of things we require.  The point is to require actions that produce an interoperable, functional protocol.  In the general case, ignoring this requirement would be bad.  At insane incoming packet rates, maybe this breaks down, and that's probably not ideal, but I find it hard to believe that you can't generate enough packets outbound to acknowledge every inbound packet at least once.  Even if you only receive every other packet, you are only spending two bytes per packet to acknowledge them.  That's almost a 600:1 advantage to the receiver.

Reply:
I think the biggest problem is if the peer assumes a packet was not processed if it eventually concludes it was not ACK'ed. That could lead to deadlocks or double spending in some poor transactional design.

Reply:
An acknowledgement indicates a packet was received and processed.  It makes no claim that other packets are never processed.  This is true throughout the QUIC design.

Reply:
The MUST de facto makes that claim, although timing can make it hard to decide.

Reply:
No, it doesn't, because you have no guarantee that you received every ACK the peer sent.  This requires the peer to put every packet into an ACK frame sent in at least one packet; there's no requirement to ensure that it's delivered reliably to the peer.  Thus, if you haven't received an ACK, that doesn't mean the peer didn't received the packet anyway.  It could be delayed and still in transit, or the ACK might have been lost.

That said, there's no actual difference between "MUST ack every packet, but not reliably" and "SHOULD ack every packet" from the other side's point of view.  This is something that could be bench-tested on an implementation, but can't be ascertained about the peer over a lossy network.

Reply:
True, I forgot that a specific PN can be present in one ACK frame on not in the next, even if there are both older and younger PNs in the next ACK frame. This makes it inconclusive.